### PR TITLE
Added Catch Primitive Class

### PR DESCRIPTION
**Pull request by matthewberends**
### Description
Added the catch primitive class.

### Testing Done
Running catkin_make builds the code / AI.
 
### Resolved Issues
#125 

### Review Checklist
***It is the reviewers responsibility to also make sure every item here has been covered***
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/thunderbots/software/ai/primitive/catch_primative.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include "primitive.h"
+
+class CatchPrimitive : public Primitive
+{
+public:
+    static const std::string PRIMITIVE_NAME;
+
+    /**
+     *
+     * Creates a new Catch Primative
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/catch_primative.h]: It's helpful to also have a brief description of what the Primitive does. Eg "Moves the robot into the ball's trajectory, using to dribbler to catch the ball and gain control, provided the robot is already in the general direction of the ball's travel"

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Just one small comment (to add a comment), but otherwise looks great! Looks like you'll need to run the `fix_formatting.sh` script too for CI to pass.

File: src/thunderbots/software/ai/primitive/catch_primitive.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#include "ai/primitive/catch_primative.h"
+
+const std::string CatchPrimitive::PRIMITIVE_NAME = "Catch Primitive";
+
+CatchPrimitive::CatchPrimitive(unsigned int robot_id, double velocity,
+                               double dribbler_speed,  double margin)
+                               : robot_id(robot_id)
+                               , velocity(velocity)
```
Comment (henryymliu) [inline src/thunderbots/software/ai/primitive/catch_primitive.cpp]: nit: would be preferable if commas were at the end of the line

**Review by henryymliu — CHANGES_REQUESTED** [review]
One minor nit from my end, otherwise looks good.

File: src/thunderbots/software/ai/primitive/catch_primative.h (hunk 1 above)
Comment (EvanMorcom) [inline src/thunderbots/software/ai/primitive/catch_primative.h]: Nit: Primitive spelling

**Review by EvanMorcom — COMMENTED** [review]
After addressing the current comments, LGTM 👍

File: src/thunderbots/software/ai/primitive/catch_primitive.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/catch_primitive.cpp]: I agree with Henry.

**Comment by MathewMacDougall** [comment]
@matthewberends it would be great if you could also add some unit tests for this class as well. That will give us test coverage right away, and is a good introduction to testing in our codebase.

You can use #131 as an example. Sorry for the last minute change.

File: src/thunderbots/software/ai/primitive/catch_primitive.cpp (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/primitive/catch_primitive.cpp]: Not sure about this, but this style might be enforced by clang-format. Just a heads up. (If it is, we can open a separate PR to deal with it)

Comment (henryymliu) [reply src/thunderbots/software/ai/primitive/catch_primitive.cpp]: Ah, just noticed this failed clang-format. @matthewberends can you run fix_formatting.sh from the clang_format folder?

File: src/thunderbots/software/ai/primitive/catch_primative.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+     *
+     * @param robot_id The id of the Robot to run this Primitive
+     * @param velocity Velocity to move robot forwards/backwards
+     * to catch the ball without it bouncing off the dribbler; units are m/s
+     * @param dribbler_speed Speed to rotate dribbler at. Units are RPM.
+     * @param margin THe margin of allowed error in the robot catch velocity.
+     * The exact use of this parameter needs to be confirmed.
```
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/catch_primative.h]: @MathewMacDougall - can we confirm what this is used (or not used) for?

@matthewberends - if we're not using this, we should be able to just remove it (we can always add it back later)

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/ai/primitive/catch_primative.h (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/catch_primative.h]: Confirmed it is used, and is a scaling factor for how far in front of the ball to make the point of intercept. It scales a value based on some kind of difference in velocity between the ball and the robot (not immediately obvious how/why that relative velocity was chosen)

Eg. When the robot is in the position to catch the ball (on the ball's line of travel, in front of it, facing the ball), the ball should be `margin * magic number` meters away.

**Review by henryymliu — APPROVED** [review]
LGTM

**Review by MathewMacDougall — APPROVED** [review]
LGTM :tada: Nice work!

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**